### DISCUSSION TITLE
REQ: TTF in menus and more

### DISCUSSION THREAD

**Discussion by mushis** [ezQuake/ezquake-source#500] [Ideas]
Hi, 

True Type Fonts, love them. 
I wish the menus also supported them. they will look great (cough cough AfterQuake).
my experience using TTF so far:

- it's a pain to manually have to set every *proportional cvar. how about having one variable that would change all *proporcional cvars to 1/0? then manually tweak the ones that we want to set differently? would be less time-consuming.
- I'de also like the different cvars involved with this had a better description, mentioning each other.
- I had to tweak the values of con_width and con_height, so the size of the characters would be bigger. I don't know if that's the only way, but FTE has con_scale, which is a better way of doing this.

**Comment by mushis** [comment]
Hemostick has a great looking Hud using the fonts. check it out if you haven't. here's an example https://www.twitch.tv/videos/812089930

**Comment by Calinou** [comment]
How do I use TTF fonts in the first place? I've seen the TTF and font variables and have set `font_facepath` to a valid TTF font path, but I still have the bitmap font displaying.

My `sys_fontsdir` is `/home/hugo/.fonts` and I've set `font_facepath` to `Hack-Bold.ttf` (I have a TTF at `/home/hugo/.fonts/Hack-Bold.ttf`). If I specify an invalid font path, I get an error message, but here I'm not getting any error message.

For the record, my ezQuake version is the latest Git commit (https://github.com/ezQuake/ezquake-source/commit/e4472cf1cf37ada095e8df9637c789646cf9145a) which I just recompiled.

**Comment by paddern** [comment]
> it's a pain to manually have to set every *proportional cvar. how about having one variable that would change all *proporcional cvars to 1/0? then manually tweak the ones that we want to set differently? would be less time-consuming.

Use could use `/toggle_re proprotional$`

> I had to tweak the values of con_width and con_height, so the size of the characters would be bigger. I don't know if that's the only way, but FTE has con_scale, which is a better way of doing this.

`vid_conscale`

**Comment by hemostx** [comment]
One thing I realized I missed when setting up is an extra adjustment variable for the font size.
At a given point size, all typefaces don't quite line up evenly : when cycling through potential fonts to play with, it would be very handy to have something to adjust glyphs globally without also scaling the whole hud like vid_conscale does, or manually adjusting every element with hud_*something*_size (because you then find out that font is not going to work out, and you need to repeat dozens of micro adjustments...)

**Comment by paddern** [comment]
Small caps would also be great but I think they are a pita to implement. :)

**Comment by mushis** [comment]
> /toggle_re proprotional$

this actually worked! cool. the $ is a wildcard, right?

> vid_conscale


but this didn't. this only works when conwidth and conheight are set to 0 afaik

**Comment by meag** [comment]
vid_conscale

but this didn't. this only works when conwidth and conheight are set to 0 afaik

^ This has come up before, I don't understand... what do you think conwidth & conheight do, if not scale the size of the font?  Are you using it to change your aspect ratio in-game because that's linked to conwidth & conheight because viewsize < 100 is linked to it in swquake?

**Comment by mushis** [comment]
I use a custom vid_conwidth and vid_conheight. 
I wanted to use vid_conscale to scale the size of the font. otherwise, the only way to do it is with the above commands.

**Comment by meag** [comment]
right, but why not adjust vid_conwidth & vid_conheight further, to scale the size of the font?  what is it that makes you want to keep your existing custom vid_conwidth & conheight settings, but adjust the size of the font with something else?

**Comment by mushis** [comment]
> right, but why not adjust vid_conwidth & vid_conheight further, to scale the size of the font?

for convenience I guess. its a pita to maintain a proportion i feel comfortable with when making adjustments (which is not often) (and since I'm used to play with a custom proportion, more "stretched"). With something like vid_conscale i would set the proportion in the other variables, and set the scale in that. 

One word to reinforce the main embodiment of this issue, TTF in menus would be great. FTE has a good implementation (hint hint). it allows a different font for the main menu too.

**Comment by meag** [comment]
`since I'm used to play with a custom proportion, more "stretched"`

You've deliberately stretched your font so width/height is different, or you've stretched your 3d view?  We get these requests quite a bit, I still think that's a legacy of to software quake.exe and we should fix that.  vid_conwidth/height should be fontsize only (but isn't)

**Comment by mushis** [comment]
i've deliberately stretched my font. Others do it also. edit: its mainly to change the aspect ratio of the 3d render. example:
![ezquake050](https://user-images.githubusercontent.com/3011167/101055693-12a8b900-3582-11eb-8251-9f7463831536.png) 
I can explain the custom stuff i got there if you want :)


> I still think that's a legacy of to software quake.exe and we should fix that. vid_conwidth/height should be fontsize only (but isn't)


this is the crucial point here. Imo i don't think the current behavior of _vid_conwidth/height_ should be changed though. that will break ppl's configs, originate whine, and in the end it's a change that no other engine has afaik. So I'm against it. What I'm for, is exactly what you said: "_something_ should change fontsize only". that's exactly what I suggest _vid_conscale_ to do. change fontsize, even if _vid_conwidth/height_ aren't set to 0, respecting the proportions set in _vid_conwidth/height_.

I hope I'm clear, my English is getting worse as years come by.

**Comment by meag** [comment]
Your english is great mushi, as ever.  It's only config whine that has stopped this being solved already, but it comes up so often we'll eventually have to bite the bullet.

Unfortunately setting vid_conscale to do as you suggest won't fix it, because then players would have no way to set their 3d proportions and maintain a 1:1 font ratio, which is what we've been asked for before by foogs.  I think config whine is the only way forward, but we'll probably have to add a flag with the default being the broken behaviour, in the same way we added scr_fovmode.
